### ISSUE TITLE
News Classification

### ISSUE THREAD

**Issue by zettlingzettel** [zettlingzettel/News-Classification#1]
#Drive mount

from google.colab import drive
drive.mount('/content/drive')

#Install & Import Libraries

import os
import pandas as pd
import numpy as np

import pandas as pd
!pip uninstall keras
!pip install tensorflow transformers==4.36.0 ktrain

!pip install ktrain
import ktrain
from ktrain import text

import tensorflow as tf
from tensorflow import keras

os.environ['TF_USE_LEGACY_KERAS'] = '1'

#Set environment & define directories

ROOT_DIR = "/content/drive/MyDrive/Text_marketing"
DATA_DIR = "%s" % ROOT_DIR
EVAL_DIR = "%s/evaluation" % ROOT_DIR
MODEL_DIR = "%s/models" % ROOT_DIR

#Reading data

reviews = pd.read_json('%s/training_data.json' % DATA_DIR)

# View Dataframe Summary
reviews.info()

#Preview dataset rows

reviews.head()

#Prepare combined text firelds
reviews['combined_text'] = reviews['headline'] + ' ' + reviews['short_description']

 # Extract articles by category
reviews[reviews['category'].str.contains("HEALTHY LIVING|WELLNESS")]

# Label data for binary classification
reviews['healthy_wellness'] = np.where((reviews['category'] == 'HEALTHY LIVING') | (reviews['category'] == 'WELLNESS'), 1, 0)

# Verify health & wellness filtering
reviews[reviews['category'].str.contains("HEALTHY LIVING|WELLNESS")]

# Preview health & wellness filtering
reviews['healthy_wellness'].describe()

# Balance the datasheet
sample_amount = 20000
healthy_wellness = reviews[reviews['healthy_wellness'] == 1]
not_healthy_wellness = reviews[reviews['healthy_wellness'] == 0].sample(n = sample_amount, random_state = 1)

review_sample = pd.concat([healthy_wellness, not_healthy_wellness])

review_sample

# Split dataset & preprocess text
from sklearn.model_selection import train_test_split
from ktrain import text
from ktrain.text import texts_from_df

train, val, preprocess = texts_from_df(
    review_sample, 'combined_text',
    label_columns=['healthy_wellness'],
    val_pct=0.1,
    maxlen=32,
    max_features=7000,
    ngram_range=1,
    preprocess_mode='standard',
    verbose=1,
    random_state=1
    )

# Build model & learner
model = text.text_classifier('fasttext', train_data=train, preproc=preprocess)
learner = ktrain.get_learner(model, train_data=train, val_data=val, batch_size=32)

# Find learning rate
learner.lr_find(max_epochs=6)

learner.lr_plot()

# Train the model
history = learner.autofit(1e-4, checkpoint_folder = 'checkpoint', epochs = 16, early_stopping = True)

# Save model
learner.save_model("%s/distilbertmodel_learner" % DATA_DIR)

# Create predictor & save
predictor = ktrain.get_predictor(learner.model, preproc=preprocess)
predictor.save("%s/distilbertmodel_predictor")

#Test Documents used in Lecture

test_docs = [
'Stress May Be Your Heart’s Worst Enemy Psychological stress activates the fear center in the brain, setting into motion a cascade of reactions that can lead to heart attacks and strokes.',
'Exercising to Slim Down? Try Getting Bigger. It’s high time for women to reclaim the real strength behind exercise.',
'What Are Your Food Resolutions for the New Year? Join us for the Eat Well Challenge starting in January.',
'Why We All Need to Have More Fun. Prioritizing fun may feel impossible right now. But this four-step plan will help you rediscover how to feel more alive.',
'Cuomo Will Not Be Prosecuted in Groping Case, Albany D.A. Says. The district attorney described the woman who said former Gov. Andrew Cuomo had groped her as “credible,” but added that proving her allegation would be difficult.',
'A Film Captures Jewish Life in a Polish Town Before the Nazis Arrived. A documentary based on a home movie shot by an American in 1938 provides a look at the vibrancy of a Jewish community in Europe just before the Holocaust.'
             ]

for i, text in enumerate(test_docs):
  probs = predictor.predict(text, return_proba = True)
  print("---------------------------")
  print('The probability this is health or wellness is %s' % probs[1])
  print(text)

validation = learner.validate(val_data=val, print_report=True)
